incubator/hnc/internal/validators/namespace.go

```diff
@@ -64,48 +64,70 @@ func (v *Namespace) Handle(ctx context.Context, req admission.Request) admission
 // handle implements the non-boilerplate logic of this validator, allowing it to be more easily unit
 // tested (ie without constructing a full admission.Request).
 func (v *Namespace) handle(req *nsRequest) admission.Response {
-	parent := req.ns.Annotations[api.SubnamespaceOf]
-	if parent != "" && req.op == v1beta1.Delete {
-		msg := fmt.Sprintf("The namespace %s is a subnamespace. Please delete the anchor from the parent namespace %s to delete the subnamespace.", req.ns.Name, parent)
-		return deny(metav1.StatusReasonForbidden, msg)
-	}
-
-	return v.checkForest(req)
-}
-
-func (v *Namespace) checkForest(req *nsRequest) admission.Response {
 	v.Forest.Lock()
 	defer v.Forest.Unlock()
 
 	ns := v.Forest.Get(req.ns.Name)
 
-	// Allow all create namespace requests except if the namespace name already exists in
-	// the external hierarchy (the apiserver already denies existing namespace names).
-	if req.op == v1beta1.Create {
-		for _, nm := range v.Forest.GetNamespaceNames() {
-			if _, ok := v.Forest.Get(nm).ExternalTreeLabels[req.ns.Name]; ok {
-				msg := fmt.Sprintf("The namespace name %q is reserved by the external hierarchy manager %q.", req.ns.Name, v.Forest.Get(nm).Manager)
-				return deny(metav1.StatusReasonAlreadyExists, msg)
-			}
+	switch req.op {
+	case v1beta1.Create:
+		// Allow all create namespace requests except if the namespace name already exists in
+		// the external hierarchy (the apiserver already denies existing namespace names).
+		return v.nameExistsInExternalHierarchy(req)
```

Comment:
Nit: TBH, even though it's a _bit_ more verbose, I'd actually suggest using the same pattern as on line 82 for _all_ these test, and then have a single `allow("")` at the end of the function. That makes it easier to add new tests in the future and also, I think, reads a bit better as English.

For example, it looks to me like `nameExistsInExternalHierarchy()` should return `true` if the name already exists, which means... that the request should be allowed? That's not right. But maybe if we call it:

```
if rsp := v.conflictsWithExternalHierarchy(req); !rsp.Allowed {
   return rsp
}
```

to me, that reads more clearly even though it's more verbose (which is usually the opposite of what I like).

Anyway, I don't actually feel too strongly about this, so up to you if you want to make this change or not.

Reply:
Agreed on both new tests in the future and easy to read!